### PR TITLE
Add AuCPace protocol to PAKEs

### PR DESCRIPTION
**Pull request by tritoke**
AuCPace is modern V-PAKE designed for Industrial IOT applications, it's inclusion in RustCrypto/PAKEs should promote the general use of PAKEs and enable their use for microcontrollers.

### CODE REVIEW

**Comment by tarcieri** [comment]
Looks like you'll need a rebase and to bump MSRV to 1.60

**Comment by tritoke** [comment]
Okay this seems to be a problem with password_hash switching to using CryptoRngCore instead of RngCore + CryptoRng.

**Comment by tritoke** [comment]
got it building now so I'm gonna sneak it into the original commit and add the MSRV bump back on top

**Comment by tritoke** [comment]
Okay final™ failure is clippy complaining about an unnecessary `.to_vec()` so I'm just gonna sneak that in the MSRV bump commit.

**Comment by tarcieri** [comment]
> It seems like I might have to bump spake2 and srp to 1.60 as well as I use dep: :(

That's fine. 1.60 is a difficult MSRV bump, especially for workspaces

**Comment by tritoke** [comment]
Okay all passed, an extra commit snuck in there so I've squashed that into the Add AuCPace commit :)
gonna pretend that like 10 force commits just didn't happen 😂

File: aucpace/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+exclude = [".gitignore"]
+readme = "README.md"
+edition = "2021"
+rust-version = "1.60"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+curve25519-dalek = { version = "4.0.0-rc.1", default-features = false, features = ["digest", "rand_core"] }
+password-hash = { version = "0.4", default-features = false, features = ["rand_core"] }
```
Comment (tarcieri) [inline aucpace/Cargo.toml:21]: This is out-of-date
```suggestion
password-hash = { version = "0.5", default-features = false, features = ["rand_core"] }
```

File: aucpace/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+rand_core = { version = "0.6", default-features = false }
+serde = { version = "1", default-features = false, optional = true, features = ["derive"] }
+serde_arrays = { version = "0.1", optional = true }
+subtle = { version = "2.4", default-features = false }
+scrypt = { version = "0.10", default-features = false, optional = true, features = ["simple"] }
```
Comment (tarcieri) [inline aucpace/Cargo.toml:26]: This is also out-of-date
```suggestion
scrypt = { version = "0.11", default-features = false, optional = true, features = ["simple"] }
```

File: aucpace/Cargo.toml (hunk 1 above)
Comment (tritoke) [reply aucpace/Cargo.toml:21]: Thanks I'll add a commit to bump these two dependencies.

File: aucpace/examples/key_agreement.rs [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+use password_hash::{ParamsString, SaltString};
+use rand_core::OsRng;
+use scrypt::{Params, Scrypt};
+use sha2::digest::Output;
+use sha2::Sha512;
+use std::io::{Read, Write};
+use std::net::{IpAddr, Ipv4Addr, SocketAddr, TcpListener, TcpStream};
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::thread;
+use std::time::Instant;
```
Comment (tarcieri) [inline aucpace/examples/key_agreement.rs:12]: Most of our other crates support `no_std`, in most cases with heapless functionality.

It's fine if `alloc` is a hard dependency, but would it be possible to make the `std`-dependent features optional?

**Comment by tritoke** [comment]
So I've found an issue in that the crate cannot actually compile for no_std targets with serde because serde_arrays isn't a no_std library, theres no reason for it not to be so I'm drafting a PR to make it no_std compatible. As a replacement serde-byte-array is no_std and provides what we need as [u8; N] is the only type I needed to serialise with serde-arrays.

Should I add another commit on top or just hide it inside of the one before the version bump again?

**Comment by tritoke** [comment]
Should have looked at the serde_arrays PRs... one has been open for two years about adding no_std support and it has a full implementation so I'm just going to leave it.

**Comment by tritoke** [comment]
For now I've just added a commit on top but I can squash it if need be.

**Comment by tarcieri** [comment]
Does `serde_bytes` do what you need? It's `no_std`-compatible and maintained by the `serde` team.

**Comment by tritoke** [comment]
serde_bytes doesn't seem to support arrays from what I can tell, only Vec<u8> and &[u8]. serdect I think would be usable, though it would require some modification as serde-byte-array just gives me a serde_with helper that works. I'm also unsure of the benefit of a constant time implementation, the only "secret" value that a message can ever hold is the Registration message, which contains a verifier. I'm going to head to bed now but I will take a look at serdect tomorrow.

**Comment by tritoke** [comment]
So I've had a look at serdect and I can't see a good reason to switch to it, I would have to manually implement Serialize and Deserialize, the only sensitive value that could be leaked is a verifier, which is only sent once per user as it is part of the registration packet, I thus think the risk is minimal from this attack. I'm going to stick with serde-byte-array, fwiw this crate is also maintained by people doing cryptography on embedded hardware.

**Pull request merged**